### PR TITLE
ENH/PERF: Remove frequency inference from .dt accessor

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: pandas/tests/series/test_datetime_values.py [hunk 1]
```diff
@@ -127,8 +127,14 @@ def compare(s, name):
...
 
         # datetimeindex with tz
-        s = Series(date_range('20130101', periods=5, tz='US/Eastern'),
-                   name='xxx')
+        values = [
+            '2012-01-01 13:00:00',
+            '2012-01-01 12:01:00',
+            '2012-01-01 08:00:00'
+        ]
+        s = Series(pd.to_datetime(values).values, name='xxx')
```
Comment (jreback) [inline pandas/tests/series/test_datetime_values.py]: you don't need the ``.values``

**Comment by jreback** [comment]
can you run the asv's with --bench freq and see what improves.

**Comment by jreback** [comment]
also pls add a whatsnew note (perf section)

**Comment by jorisvandenbossche** [comment]
@cpcloud can you give a bit more context? 
What is the consequence of removing it? Why what the inference added in the first place?

I suppose the only reason to have the `freq='infer'` when constructing the DatetimeIndex for the `dt` accessor is for the `dt.freq` attribute? (I don't think the other methods available in `dt.` need the freq) 
But removing the `freq='infer'` makes that `dt.freq` will always be None and thus rendered effectively useless. And it *is* also a backwards incompatible change (so needs a note about that in the whatsnew). If we do it like this, I think we should remove (or deprecate) `dt.freq`

But maybe we can also do the infer step only when people ask for `dt.freq` ? (so there is no backwards incompatible change) Not sure how complicated it would be, because then it doesn't map directly to get the `freq` from the underlying datetimeindex object that was constructed

File: pandas/tests/series/test_datetime_values.py [hunk 2]
```diff
@@ -127,8 +127,14 @@ def compare(s, name):
         tm.assert_series_equal(result, expected)
 
         # datetimeindex with tz
-        s = Series(date_range('20130101', periods=5, tz='US/Eastern'),
-                   name='xxx')
+        values = [
+            '2012-01-01 13:00:00',
```
Comment (jorisvandenbossche) [inline pandas/tests/series/test_datetime_values.py]: is there a reason you change those values?

**Comment by jreback** [comment]
this was removed for perf, IOW we *don't* need to infer by default (we don't normally do this on construction). so the ``.freq`` attribute in the accessor could simply return ``inferred_freq`` instead (which does the inference if needed).

File: pandas/tests/series/test_datetime_values.py (hunk 2 above)
Comment (cpcloud) [inline pandas/tests/series/test_datetime_values.py]: The original values are already given a frequency by constructing with `date_range` so the test that was confirming their frequency wasn't actually doing anything useful.

**Comment by cpcloud** [comment]
@jorisvandenbossche the inference on attribute access is suboptimal for two reasons. 1. It makes attribute access scale with the size of the series. 2. Any order N operation on the dt values is now 2*N which is noticeable for large series. We shouldn't make users pay for something they don't necessarily need and we should try to avoid leaving 2x on the table if we don't have to.

I don't think this needs to be backwards incompatible, I'll fix it up.

**Comment by jreback** [comment]
lgtm. can you add a perf note

**Comment by jreback** [comment]
thanks @cpcloud

**Pull request merged**